pkg/controller/reconciler/flux_controller.go

```diff
@@ -73,13 +78,77 @@ func deploymentTemplateRepositoryIndexer(o client.Object) []string {
 }
 
 func (r *FluxController) SetupWithManager(mgr ctrl.Manager) error {
+	r.initialized = &atomic.Bool{}
+
+	// Register a controller for CustomResourceDefinition events.
+	// We want to watch for the GitRepository CRD to be created or updated.
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&apiextensionsv1.CustomResourceDefinition{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return r.registerFluxController(e.Object, mgr)
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return r.registerFluxController(e.ObjectNew, mgr)
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return false
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				return false
+			},
+		}).
+		Complete(r)
+
+	return err
+}
+
+// registerFluxController registers the FluxController with the manager
+// when the GitRepository CRD is created or updated.
+func (r *FluxController) registerFluxController(obj client.Object, mgr ctrl.Manager) bool {
```

Comment:
Need to do all of this roundabout registering in case `GitRepository` type doesn't exist on the cluster yet. This structure means that the controller will set up a manager for `CustomResourceDefinition`, wait for `GitRepository`, and finally register the FluxController once this is all set up.